rooms.js

```diff
@@ -163,6 +165,110 @@ var Room = (function () {
+			this.unmute(this.mutedUsers[0].userid, true);
+			//runMuteTimer() is called again in unmute() so this function instance should be closed
+			return;
+		}
+		var self = this;
+		this.muteTimer = setTimeout(function () {
+			self.muteTimer = null;
+			self.runMuteTimer();
+		}, timeUntilExpire);
+	};
+	Room.prototype.isMuted = function (user) {
+		if (!user) return;
+		if (this.mutedUsers) {
+			for (var i = 0; i < this.mutedUsers.length; i++) {
+				if (user.userid === this.mutedUsers[i].userid ||
+					user.guestNum === this.mutedUsers[i].guestNum ||
+					(user.autoconfirmed && user.autoconfirmed === this.mutedUsers[i].autoconfirmed)) {
+					return this.mutedUsers[i].userid;
+				}
+			}
+		}
+	};
+	Room.prototype.getMuteTime = function (user) {
+		var userid = this.isMuted(user);
+		if (!userid) return;
+		for (var i = 0; i < this.mutedUsers.length; i++) {
+			if (userid === this.mutedUsers[i].userid) {
+				return this.mutedUsers[i].time - Date.now();
+			}
+		}
+	};
+	Room.prototype.mute = function (user, setTime) {
+		var userid = user.userid;
+
+		if (!setTime) setTime = 7 * 60000; // default time: 7 minutes
+		if (setTime < 1) setTime = 1; // mostly to prevent bugs
+		if (setTime > 90 * 60000) setTime = 90 * 60000; // limit 90 minutes
+
+		// If the user is already muted, the existing queue position for them should be removed
+		if (this.isMuted(user)) this.unmute(user);
+
+		// Place the user in a queue for the unmute timer
+		for (var queue = 0; queue < this.mutedUsers.length + 1; queue++) {
+			var timeRemaining = this.mutedUsers[queue] ? this.mutedUsers[queue].time - Date.now() : Infinity;
+			if (setTime < timeRemaining) {
+				var entry = {
+					userid: userid,
+					time: Date.now() + setTime,
+					guestNum: user.guestNum,
+					autoconfirmed: user.autoconfirmed
+				};
+				this.mutedUsers.splice(queue, 0, entry);
+				// The timer needs to be switched to the new entry if it is to be unmuted
+				// before the entry the timer is currently running for
+				if (queue === 0 && this.muteTimer) {
+					clearTimeout(this.muteTimer);
+					this.muteTimer = null;
+				}
+				break;
+			}
+		}
+		this.runMuteTimer();
+
+		user.updateIdentity(this.id);
+		return userid;
+	};
+	Room.prototype.unmute = function (userid, sendPopup) {
+		var user = Users(userid);
+		if (!user) {
+			//if the user cannot be found, construct a dummy object for them
+			user = {userid: userid};
+		}
+
+		var successUserid = false;
+		for (var i = 0; i < this.mutedUsers.length; i++) {
+			var entry = this.mutedUsers[i];
+			if (entry.userid === userid ||
+				entry.guestNum === user.guestNum ||
+				(entry.autoconfirmed && entry.autoconfirmed === user.autoconfirmed)) {
+				if (i === 0) {
+					clearTimeout(this.muteTimer);
+					this.muteTimer = null;
+					this.mutedUsers.splice(0, 1);
+					this.runMuteTimer();
+				} else {
+					this.mutedUsers.splice(i, 1);
+				}
+				successUserid = userid;
+				i--;
+			}
+		}
+
+		if (user.connected) {
```

Comment:
I'm not sure `connected` is the right thing to test here, but it seems to work how I intended.